### DISCUSSION TITLE
DLite now supports copy/paste from system clipboard (version 0.9.7)

### DISCUSSION THREAD

**Discussion by renaudpawlak** [cincheo/daquota#12] [Announcements]
This is great news because it means you can copy any DLite JSON descriptor from anywhere and paste it in your project.

This will allow great agility and flexibility for projet creators. Typically, one can provide a DLite example as a JSON, and the community can use it instantly with a simple copy/paste.

As an example, you can now copy/paste this small example of a counter that can be incremented/decremented with 2 buttons (example inspired from issue #10):

```json
{
  "layout": "block",
  "components": [
    {
      "label": "My counter",
      "inputType": "number",
      "description": "",
      "field": "",
      "size": "default",
      "disabled": false,
      "placeholder": "",
      "type": "InputView",
      "eventHandlers": [],
      "cid": "input-1",
      "defaultValue": "=1"
    },
    {
      "layout": "block",
      "components": [
        {
          "dataSource": "$parent",
          "label": "Increment",
          "buttonType": "button",
          "variant": "secondary",
          "size": "default",
          "pill": false,
          "squared": false,
          "block": false,
          "disabled": false,
          "eventHandlers": [
            {
              "global": false,
              "name": "@click",
              "actions": [
                {
                  "targetId": "input-1",
                  "name": "setData",
                  "description": "Default action",
                  "argument": "$d(target) + 1"
                }
              ]
            }
          ],
          "type": "ButtonView",
          "cid": "button-2",
          "icon": "arrow-up",
          "layoutClass": "m-2"
        },
        {
          "dataSource": "$parent",
          "label": "Decrement",
          "buttonType": "button",
          "variant": "secondary",
          "size": "default",
          "pill": false,
          "squared": false,
          "block": false,
          "disabled": false,
          "eventHandlers": [
            {
              "global": false,
              "name": "@click",
              "actions": [
                {
                  "targetId": "input-1",
                  "name": "setData",
                  "description": "Default action",
                  "argument": "$d(target) - 1"
                }
              ]
            }
          ],
          "type": "ButtonView",
          "cid": "button-3",
          "icon": "arrow-down",
          "layoutClass": "m-2"
        }
      ],
      "defaultValue": "={}",
      "type": "ContainerView",
      "eventHandlers": [],
      "cid": "container-1",
      "direction": "row",
      "justify": "center"
    }
  ],
  "defaultValue": "={}",
  "type": "ContainerView",
  "eventHandlers": [],
  "cid": "container-2"
}
```

The only drawback I can see would be navigator support for the clipboard API (your need to allow the browser to read the clipboard at some point). Tested under Chrome, and it works perfectly. Not sure that DLite works fine under IE anyways :)

**Comment by HFolkertH** [comment]
hello Renauld,

Nice new feature!
But where should I paste the Json code?

**Reply by renaudpawlak** [reply]
You select a place where you want to paste the component. To do so, move your mouse over the different places and when you see a highlight, you click on it. It should become bigger with a "plus" icon. Now you can copy and the clipboard content (if a valid dlite model) will insert at this place.

**Reply by HFolkertH** [reply]
Hello Renaud,

I am using the newest Chrome on my MacBook air (Big Sur).  I also disabled AdBlocker.
I copied the code above, then clicked the big Plus in a new Container  but pasting (with keyboard-shortcut and with Chrome/edit/paste menu) does not work. Is it a configuration in Chrome. Any tips?

**Reply by HFolkertH** [reply]
Perhaps use the Edit menu in the IDE?  Add a "Paste code" ?

**Reply by renaudpawlak** [reply]
That's strange. I have a Mac Book Pro with Big Sur also, and I also use the latest chrome. Normally "Command + V" should work... it works with me. You version of DLite is 0.9.6 right?

**Reply by renaudpawlak** [reply]
@HFolkertH my bad, you were right it did not work well. I have found the problem and fixed it. It should work now (version 0.9.7). Thank you for your help.

**Comment by HFolkertH** [comment]
I just discovered in the left side Builder menu the Generic item. It opens a JSON editarea where you can next paste the code (ctrl-v). After clicking OK the canvas shows all the components. Easy when you know it.

**Reply by renaudpawlak** [reply]
Good catch! That's exactly the same as a paste :) (more steps, but it also allows you to change the JSON if needed)

**Comment by hasen666** [comment]
> This is great news because it means you can copy any DLite JSON descriptor from anywhere and paste it in your project.
> 
> This will allow great agility and flexibility for projet creators. Typically, one can provide a DLite example as a JSON, and the community can use it instantly with a simple copy/paste.
> 
> As an example, you can now copy/paste this small example of a counter that can be incremented/decremented with 2 buttons (example inspired from issue #10):
> 
> ```json
> {
>   "layout": "block",
>   "components": [
>     {
>       "label": "My counter",
>       "inputType": "number",
>       "description": "",
>       "field": "",
>       "size": "default",
>       "disabled": false,
>       "placeholder": "",
>       "type": "InputView",
>       "eventHandlers": [],
>       "cid": "input-1",
>       "defaultValue": "=1"
>     },
>     {
>       "layout": "block",
>       "components": [
>         {
>           "dataSource": "$parent",
>           "label": "Increment",
>           "buttonType": "button",
>           "variant": "secondary",
>           "size": "default",
>           "pill": false,
>           "squared": false,
>           "block": false,
>           "disabled": false,
>           "eventHandlers": [
>             {
>               "global": false,
>               "name": "@click",
>               "actions": [
>                 {
>                   "targetId": "input-1",
>                   "name": "setData",
>                   "description": "Default action",
>                   "argument": "$d(target) + 1"
>                 }
>               ]
>             }
>           ],
>           "type": "ButtonView",
>           "cid": "button-2",
>           "icon": "arrow-up",
>           "layoutClass": "m-2"
>         },
>         {
>           "dataSource": "$parent",
>           "label": "Decrement",
>           "buttonType": "button",
>           "variant": "secondary",
>           "size": "default",
>           "pill": false,
>           "squared": false,
>           "block": false,
>           "disabled": false,
>           "eventHandlers": [
>             {
>               "global": false,
>               "name": "@click",
>               "actions": [
>                 {
>                   "targetId": "input-1",
>                   "name": "setData",
>                   "description": "Default action",
>                   "argument": "$d(target) - 1"
>                 }
>               ]
>             }
>           ],
>           "type": "ButtonView",
>           "cid": "button-3",
>           "icon": "arrow-down",
>           "layoutClass": "m-2"
>         }
>       ],
>       "defaultValue": "={}",
>       "type": "ContainerView",
>       "eventHandlers": [],
>       "cid": "container-1",
>       "direction": "row",
>       "justify": "center"
>     }
>   ],
>   "defaultValue": "={}",
>   "type": "ContainerView",
>   "eventHandlers": [],
>   "cid": "container-2"
> }
> ```
> 
> The only drawback I can see would be navigator support for the clipboard API (your need to allow the browser to read the clipboard at some point). Tested under Chrome, and it works perfectly. Not sure that DLite works fine under IE anyways :)

This is a very interesting feature !
